File: server/src/utils/ws_event.ts
export class WSEvent {
  eventName: string;
  recieverUsername: string;
  senderUsername: string;
  data: string | undefined;
  time: Number;

  constructor(
    eventName: string,
    recieverUsername: string,
    senderUsername: string,
    data: string | undefined,
    time: Number
  ) {
    this.eventName = eventName;
    this.recieverUsername = recieverUsername;
    this.senderUsername = senderUsername;
    this.data = data;
    this.time = time;
  }
  static fromJson(jsonObject: string): WSEvent {
    var map = JSON.parse(jsonObject);
    return new WSEvent(
      map["eventName"],
      map["recieverUsername"],
      map["senderUsername"],
      map["data"],
      map["time"]
    );
  }
  ///used
  toJson(): string {
    return JSON.stringify({
      eventName: this.eventName,
      recieverUsername: this.recieverUsername,
      senderUsername: this.senderUsername,
      data: this.data,
      time: this.time,
    });
  }
}

File: client/lib/core/helper/websocket/ws_event.dart
import 'dart:convert';

import 'package:client/common/entity/message.dart';

// ignore_for_file: public_member_api_docs, sort_constructors_first
class WSEvent {
  final String eventName;
  final String senderUsername;
  final String recieverUsername;
  final String? data;
  final DateTime time;

  const WSEvent(
    this.eventName,
    this.senderUsername,
    this.recieverUsername,
    this.data,
    this.time,
  );

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'eventName': eventName,
      'senderUsername': senderUsername,
      'recieverUsername': recieverUsername,
      'data': data,
      'time': time.millisecondsSinceEpoch,
    };
  }

  Message toMessage() {
    return Message(
      chatId: null,
      content: data!,
      isme: false,
      time: time,
    );
  }

  factory WSEvent.fromMap(Map<String, dynamic> map) {
    return WSEvent(
      map['eventName'] as String,
      map['senderUsername'] as String,
      map['recieverUsername'] as String,
      map['data'] as String?,
      DateTime.fromMillisecondsSinceEpoch(map['time']),
    );
  }

  String toJson() => json.encode(toMap());

  factory WSEvent.fromJson(String source) =>
      WSEvent.fromMap(json.decode(source) as Map<String, dynamic>);
}
